### PR TITLE
parse TernaryIf as right associative, fix #68221

### LINKED ISSUE #68221: [JIT][C++ Parser] Bug in processing ternary expressions
## 🐛 Bug 

I ran into the following bug when specifying shape functions in C++. Because we rely on the python ast parser generally, and then export ternary expressions as if statements, this doesn't come up very often. 

```
import torch

cu = torch.jit.CompilationUnit('''
    def foo(x: List[int]):
        return 1 if len(x) == 0 else x[0] if len(x) == 1 else x[1]
''')
print(cu.foo([]))
```
throws an exception

### PR DESCRIPTION
**Pull request by etern**
Fixes #68221

### CODE REVIEW

**Comment by eellison** [comment]
thanks for pr!! do you mind adding tests ?

**Comment by etern** [comment]
I add test case to compile this code snippet:
```python
def plus_123(x: int):
  return x + 1 if x == 1 else x + 2 if x == 2 else x + 3
```
Before the fix `print(plus_123.code)`:
```python
def plus_123(x: int) -> int:
  if torch.eq(x, 2):
    if torch.eq(x, 1):
      _1 = torch.add(x, 1)
    else:
      _1 = torch.add(x, 2)
    _0 = _1
  else:
    _0 = torch.add(x, 3)
  return _0
```
After the fix `print(plus_123.code)`:
```python
def plus_123(x: int) -> int:
  if torch.eq(x, 1):
    _0 = torch.add(x, 1)
  else:
    if torch.eq(x, 2):
      _1 = torch.add(x, 2)
    else:
      _1 = torch.add(x, 3)
    _0 = _1
  return _0
```

File: torch/csrc/jit/frontend/lexer.h [hunk 1]
```diff
@@ -286,6 +286,7 @@ struct TORCH_API SharedParserData {
     switch (kind) {
       case '?':
       case TK_POW:
```
Comment (tugsbayasgalan) [inline torch/csrc/jit/frontend/lexer.h:288]: should this return True too?

File: test/cpp/api/jit.cpp [hunk 2]
```diff
@@ -125,3 +125,13 @@ TEST(TorchScriptTest, TestPickle) {
   double eps = 0.0001;
   ASSERT_TRUE(diff < eps && diff > -eps);
 }
+
+TEST(TorchScriptTest, TestTernaryIfRightAssociative) {
```
Comment (tugsbayasgalan) [inline test/cpp/api/jit.cpp:129]: Can you make it a python test case under test/test_jit.py?

File: torch/csrc/jit/frontend/lexer.h (hunk 1 above)
Comment (etern) [reply torch/csrc/jit/frontend/lexer.h:288]: Sorry I'm not sure about your question, did you mean `TK_POW`？It do returns true, the switch-case will fall through

My fix is add true case for `TK_IF`, other case is not changed

File: test/cpp/api/jit.cpp (hunk 2 above)
Comment (etern) [reply test/cpp/api/jit.cpp]: Yes, moved test case to test_jit.py.

..and I wonder, a little, why test_jit.py is better than jit.cpp?

File: torch/csrc/jit/frontend/lexer.h (hunk 1 above)
Comment (tugsbayasgalan) [reply torch/csrc/jit/frontend/lexer.h:288]: Isn't TK_POW return true in the original statement?

File: test/test_jit.py [hunk 3]
```diff
@@ -6579,6 +6579,15 @@ def forward(self) -> torch.Tensor:
         self.assertEqual(model1.forward(), script_model_1.forward())
         self.assertEqual(model2.forward(), script_model_2.forward())
 
+    def test_ternary_right_associative(self):
```
Comment (tugsbayasgalan) [inline test/test_jit.py:6582]: I think you can do sth like:
```
    def test_ternary_right_associative(self):
        @torch.jit.script 
        def plus_123(x: int):
               return x + 1 if x == 1 else x + 2 if x == 2 else x + 3
        self.checkScript(plus_123, (2,))
        ....

```

File: test/cpp/api/jit.cpp (hunk 2 above)
Comment (tugsbayasgalan) [reply test/cpp/api/jit.cpp]: The reason is the frontend parsing is done in python for test_jit.py which is more robust than our own parser. We rely on Python AST if the code is written in python first. That being said, I don't think it really matters in your case, just a convention really

File: test/test_jit.py (hunk 3 above)
Comment (etern) [reply test/test_jit.py:6582]: 👌 changed to self.checkScript

File: torch/csrc/jit/frontend/lexer.h (hunk 1 above)
Comment (etern) [reply torch/csrc/jit/frontend/lexer.h:288]: TK_POW should be right associative `x**2**3 == x**(2**3)`
it return true in the original statement
it return true in my PR
I didn't changed it

**Review by tugsbayasgalan — APPROVED** [review]

**Comment by eellison** [comment]
@tugsbayasgalan mind landing?

**Comment by eellison** [comment]
Hi @etern, sorry about the delay here, do you mind rebasing and i'll land ?

**Comment by eellison** [comment]
yea that's fine!

**Pull request closed without merging**